provisioning/device/src/tpm_registration.ts

```diff
@@ -0,0 +1,239 @@
+import { EventEmitter } from 'events';
+import * as machina from 'machina';
+import * as dbg from 'debug';
+const debug = dbg('azure-iot-provisioning-device:TpmRegistration');
+import { RegistrationClient, TpmProvisioningTransport, TpmSecurityClient, TpmRegistrationInfo } from './interfaces';
+
+export class TpmRegistration extends EventEmitter implements RegistrationClient {
+  private _fsm: machina.Fsm;
+  private _transport: TpmProvisioningTransport;
+  private _securityClient: TpmSecurityClient;
+  private _idScope: string;
+
+  constructor(idScope: string, transport: TpmProvisioningTransport, securityClient: TpmSecurityClient) {
+    super();
+    this._idScope = idScope;
+    this._transport = transport;
+    this._securityClient = securityClient;
+
+    this._fsm = new machina.Fsm({
+      namespace: 'tpm-registration',
+      initialState: 'notStarted',
+      states: {
+        notStarted: {
+          _onEnter: (err, callback) => {
+            if (callback) {
+              callback(err);
+            } else if (err) {
+              this.emit('error', err);
+            }
+          },
+          register: (callback) => this._fsm.transition('authenticating', callback),
+          cancel: (callback) => callback()
+        },
+        authenticating: {
```

Comment:
>authenticating [](start = 8, length = 14)

Should we add explicit error handling if the user calls register while a registration is in progress?  Or does the FSM provide a meaningful enough error if there's no hander available in the current state?

Reply:
the '*' action will defer it until the next transition - then the next transition takes care of it.

---
In reply to: [152068231](https://github.com/Azure/azure-iot-sdk-node/pull/155#discussion_r152068231) [](ancestors = 152068231)